Alerting: In migration, create one label per channel

JacobsonMT opened the pull request:
This PR changes how routing is done by the legacy alerting migration. 

Previously, we created a single label on each alert rule that contained an array of contact point names. Ex: `__contact__="slack legacy testing","slack legacy testing2"`

This label was then routed against a series of regex-matching policies with `continue=true`. Ex: `__contacts__ =~ .*"slack legacy testing".*`

More details here: https://github.com/grafana/grafana/pull/52071

In the case of many contact points, this array could quickly become difficult to manage and difficult to grok at-a-glance.

This PR replaces the single `__contact__` label with multiple `__legacy_c_{contactname}__` labels and simple equality-matching policies. These channel-specific policies are nested in a single route under the top-level route which matches against `__legacy_use_channels__ = true` for ease of organization.

This should improve the experience for users wanting to keep the default migrated routing strategy but who also want to modify which contact points an alert sends to.

**Special notes for your reviewer:**

![image](https://github.com/grafana/grafana/assets/8484471/cd1e1187-e32f-43b1-b2b3-7f86fe7ac9ea)

![image](https://github.com/grafana/grafana/assets/8484471/c096b712-a83e-4e02-a6ec-e1e6771ec7d5)


Please check that:
- [X] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [X] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

JacobsonMT committed:
In migration, create one label per channel using UID instead of name

Previously, we used the notification channel names to create the routing labels
on an alert rule. This makes future work to individually re-migrate
single channels difficult as channel names can be easily changed and are not
guaranteed unique by any db constraint.

This change modifies the logic to instead use the channel uid, which is
guaranteed unique per org at the db level and is significantly less likely to
change (indeed even if it does change, we can likely assume that the new channel
 is not equal in identity to the old one).

In addition, we move away from a single `__contact__` label with an array of
receiver names combined with regex-based route matching to multiple
`__contact_{uid}__` labels and a simple equality-based route matching. These
routes are now all nested under a single top-level route matched against
`__use_legacy_channels__ = true` for ease of organization.

This is done to improve the experience for users wanting to keep the
default migrated routing strategy but modify which contact points an alert
sends to. Editing the array could be difficult when large.

This has the added benefit of great simplifying the logic around contact point
migration as well and removes the (now unnecessary) DashAlert wrapper for
 legacymodels.Alert.

JacobsonMT commented:
/deploy-to-hg

yuri-tceretian commented:
Related to https://github.com/grafana/grafana/issues/56582

yuri-tceretian commented on pkg/services/ngalert/migration/channel.go:
```diff
@@ -189,197 +103,58 @@ func (om *OrgMigration) createNotifier(c *legacymodels.AlertNotification) (*apim
+	// These will match two routes as they are all defined with Continue=true.
+
+	label := fmt.Sprintf(ContactLabelTemplate, channel.UID)
+	mat, _ := labels.NewMatcher(labels.MatchEqual, label, "true")
```
`labels` is an external dependency and it's implementation details can change in future. I think it is a good practice to correctly handle the function result, in this case, error.

JacobsonMT replied:
That's fair, I went back and forth on this one. I'll add it back.

yuri-tceretian commented on pkg/services/ngalert/migration/channel.go:
```diff
@@ -180,7 +94,7 @@ func (om *OrgMigration) createNotifier(c *legacymodels.AlertNotification) (*apim
 
 	return &apimodels.PostableGrafanaReceiver{
-		UID:                   uid,
+		UID:                   c.UID,
```
I am not sure I understand what this change does. Previously, we assumed that UID of legacy notifier could be empty and generated one in that case. This change seems to discard this logic. Was the assumption incorrect? Do we want to tolerate empty UIDs?

JacobsonMT replied:
Explained below: https://github.com/grafana/grafana/pull/76527#discussion_r1364324654

yuri-tceretian commented on pkg/services/ngalert/migration/models/alertmanager.go:
```diff
@@ -0,0 +1,83 @@
+
+// UseLegacyChannelsLabel is a private label added to a rule's labels to enable routing to the nested route created
+// during migration.
+const UseLegacyChannelsLabel = "__use_legacy_channels__"
```
I would like all labels we use in alert rules to be in one place in `ngalert/models`. This will let us keep track of all system-owned labels in the system.

yuri-tceretian commented on pkg/services/ngalert/migration/models/alertmanager.go:
```diff
@@ -0,0 +1,83 @@
+type Alertmanager struct {
+	Config      *apiModels.PostableUserConfig
+	legacyRoute *apiModels.Route
+}
```
Why not use apimodels.PostableApiAlertingConfig instead and just implement some utility functions?

JacobsonMT replied:
No reason other than I found this made the code simpler. It makes a bit more sense in a future PR where `NewAlertmanager` is replaced by `FromPostableUserConfig`, but overall it just hides implementation details around where the routes are stored (and later how to clean them up).

yuri-tceretian commented on pkg/services/ngalert/migration/alert_rule.go:
```diff
@@ -287,17 +301,19 @@ func truncate(daName string, length int) string {
+			uid, err := om.migrationStore.GetAlertNotificationUidWithId(ctx, orgID, ui.ID)
+			if err != nil {
+				l.Error("Failed to get alert notification UID", "notificationId", ui.ID, "err", err)
+			}
```
I think this answers my question above: we are going to drop a notification without UID. I wonder what is the reason for such decision?

JacobsonMT replied:
If neither `id` not `uid` is present on the entry, or if the `id` is invalid, then that legacy alert would be failing to send to the channel in legacy alerting anyways.

yuri-tceretian commented on pkg/services/ngalert/migration/alert_rule.go:
```diff
@@ -6,66 +6,80 @@ import (
-	ContactLabel = "__contacts__"
+	// ContactLabelTemplate is a private label added to a rule's labels to route it to the correct migrated
+	// notification channel.
+	ContactLabelTemplate = "__contacts_%s__"
```
I think this one also should be in `ngalert/models`

yuri-tceretian replied:
I think we should add some hint to the UI page that says that such labels are not included into the notification and used only for routing purpose.

yuri-tceretian commented on pkg/services/ngalert/migration/alert_rule.go:
```diff
@@ -6,66 +6,80 @@ import (
-	annotations[ngmodels.PanelIDAnnotation] = fmt.Sprintf("%v", da.PanelID)
-	annotations["__alertId__"] = fmt.Sprintf("%v", da.ID)
+	annotations[ngmodels.PanelIDAnnotation] = fmt.Sprintf("%v", alert.PanelID)
+	annotations["__alertId__"] = fmt.Sprintf("%v", alert.ID)
```
This one and "message" below too.

yuri-tceretian commented on pkg/services/ngalert/migration/alert_rule.go:
```diff
@@ -6,66 +6,80 @@ import (
-func addMigrationInfo(da *migrationStore.DashAlert, dashboardUID string) (map[string]string, map[string]string) {
-	tagsMap := simplejson.NewFromAny(da.ParsedSettings.AlertRuleTags).MustMap()
-	lbls := make(map[string]string, len(tagsMap))
+func addLabelsAndAnnotations(l log.Logger, alert *legacymodels.Alert, dashboardUID string, channels []string) (map[string]string, map[string]string) {
```
we can use data.Labels as slightly better type for result

yuri-tceretian commented on pkg/services/ngalert/migration/alert_rule.go:
```diff
@@ -6,66 +6,80 @@ import (
-	lbls := make(map[string]string, len(tagsMap))
+func addLabelsAndAnnotations(l log.Logger, alert *legacymodels.Alert, dashboardUID string, channels []string) (map[string]string, map[string]string) {
+	tags := alert.GetTagsFromSettings()
+	lbls := make(map[string]string)
```
I think we can calculate the size of it as `len(tags) + len(channels) + 1`

yuri-tceretian commented on pkg/services/ngalert/migration/cond_trans.go:
```diff
@@ -21,11 +21,53 @@ import (
+
+// dashAlertNot is the object that represents the Notifications array in
+// dashAlertSettings
+type dashAlertNot struct {
```
Probably rename it to dashAlertNotificationKey?

yuri-tceretian commented on pkg/services/ngalert/migration/cond_trans.go:
```diff
@@ -21,11 +21,53 @@ import (
+	}
+}
+
+type conditionEvalJSON struct {
```
I wonder if we can get rid of JSON (as long as we move it anyway) suffix as it is confusing. All struct above are from JSON too.

JacobsonMT replied:
Agreed, was hoping to get rid of these but there isn't really anything suitable from legacy models without make too many changes. I'll rename some of the worst ones.

yuri-tceretian commented on pkg/services/ngalert/migration/cond_trans.go:
```diff
@@ -265,7 +307,7 @@ func getNewRefID(refIDs map[string][]int) (string, error) {
 		return sR, nil
 	}
-	return "", fmt.Errorf("failed to generate unique RefID")
+	return "", fmt.Errorf("generate unique RefID")
```
I do not think I can agree with this. This makes the chained message be more readable message `failed to migrate alert rules: failed to migrate alert rule:failed to generate unique RefID` is more readable than `migrate alert rules: migrate rule: generate unique RefID`.

JacobsonMT replied:
Yes, you're right. The leaf errors should mention the failure, this was accidental collateral in a mass replace.

JacobsonMT replied:
Though in this case it would be:

`Error "migration failed: executing migration: migrate org 1: migrate alerts: migrate and save dashboard '1': migrate alert 'alert 1': transform conditions: failed to generate unique RefID"`

yuri-tceretian commented on pkg/services/ngalert/migration/ualert.go:
```diff
@@ -4,89 +4,93 @@ import (
+			err := om.migrationStore.InsertAlertRules(ctx, rules...)
+			if err != nil {
+				return fmt.Errorf("insert alert rules: %w", err)
+			}
```
insert is moved to where it is supposed to be because now generation of rules and routes are implicitly related via the migration logic, right?

JacobsonMT replied:
Correct, before we couldn't save the alerts until we migrated the channels since they added labels. Now they are independent, so we can insert somewhere more natural.

JacobsonMT committed:
Address PR review comments

yuri-tceretian reviewed (approved):
Tested and it works as it should. LGTM

JacobsonMT committed:
Error to Warn log when notification id has no associated uid

JacobsonMT committed:
Fix small bug where folder permissions were calculated too often

JacobsonMT committed:
Use channel names in label instead of uids

JacobsonMT committed:
Modify routing labels to start with the same prefix __legacy

JacobsonMT commented:
Rebased onto main and added the following changes:

8e164e961ff0def0c585652f7dd69e270b81b867: Fixes bug where we weren't caching folder permissions correctly.
107199b7f518aa360c5be3a550e4b7b52209d21e: Use channel names instead of UIDs in routing labels.
3a9533711b193887f9d81edfe549ac5301c81f45: Rename the routing labels so they all use the same `__legacy` prefix. Should make it more obvious that they came from the legacy migration and are related.

JacobsonMT committed:
Fix rebase conflict

yuri-tceretian reviewed (approved):
Code-wise LGTM.

However, I wonder whether it does not contradict what we want to achieve as part of the simplification of notification policy management where users will be able to pick only one contact point. If we go with a different approach and create a contact point per set of the selected notification policies, then we will be able to leverage the new functionality we're adding. WDYT?

JacobsonMT committed:
Fix linting
